numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -1574,144 +1574,196 @@ _prepend_ones(PyArrayObject *arr, int nd, int ndmin, NPY_ORDER order)
                  ((order) == NPY_CORDER && PyArray_IS_C_CONTIGUOUS(op)) || \
                  ((order) == NPY_FORTRANORDER && PyArray_IS_F_CONTIGUOUS(op)))
 
-static PyObject *
-_array_fromobject(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *kws)
-{
-    PyObject *op;
-    PyArrayObject *oparr = NULL, *ret = NULL;
-    npy_bool subok = NPY_FALSE;
-    npy_bool copy = NPY_TRUE;
-    int ndmin = 0, nd;
-    PyArray_Descr *type = NULL;
-    PyArray_Descr *oldtype = NULL;
-    NPY_ORDER order = NPY_KEEPORDER;
-    int flags = 0;
+#define FAST_PARSE(string_obj, converter, value)                            \
+    do {                                                                    \
+        PyObject *tmp = PyDict_GetItem(kwargs, string_obj);                 \
+                                                                            \
+        if (kwargs_parsed == kwargs_size) {                                 \
+            goto parsed_all_arguments;                                      \
+        }                                                                   \
+                                                                            \
+        if (tmp != NULL) {                                                  \
+            kwargs_parsed++;                                                \
+            if (converter(tmp, &value) == NPY_FAIL) {                       \
+                goto finish;                                                \
+            }                                                               \
+            if (kwargs_parsed == kwargs_size) {                             \
+                goto parsed_all_arguments;                                  \
+            }                                                               \
+            num_arguments -= 1;                                             \
+            /* Make sure we expect more kwargs */                           \
+            if (num_arguments < 1) {                                        \
+                goto finish;                                                \
+            }                                                               \
+        }                                                                   \
+    } while (0)
 
-    static char *kwd[]= {"object", "dtype", "copy", "order", "subok",
-                         "ndmin", NULL};
 
-    if (PyTuple_GET_SIZE(args) > 2) {
-        PyErr_SetString(PyExc_ValueError,
-                        "only 2 non-keyword arguments accepted");
-        return NULL;
+/*
+ * Small wrapper converting to array just like CPython does, we could
+ * use our own PyArray_PyIntAsInt function, but it deprecates floats
+ * differently.
+ * Function should thus be replaced at some point.
+ */
+static
+int int_converter(PyObject *obj, int *value) {
```

Comment:
This probably works, but should maybe be moved. Also have to check if the comment is true (and it would skip the python side deprecation of coercing float to int; thus possibly breaking code on older python versions).